arch/lkl/mm/bootmem.c

```diff
@@ -12,7 +12,13 @@ void __init bootmem_init(unsigned long mem_sz)
 {
 	mem_size = mem_sz;
 
-	_memory_start = (unsigned long)lkl_ops->mem_alloc(mem_size);
+	if (lkl_ops->pci_ops && lkl_ops->pci_ops->alloc) {
+		_memory_start =
+			(unsigned long)lkl_ops->pci_ops->alloc(mem_size);
```

Comment:
Actually, pci_ops->alloc() just call mmap inside, so I updated the patch and introduced a new interface lkl_ops->page_alloc(), whose implementation is equivalent to that of pci_ops->alloc().

Now, PCI-related functions are removed from kernel memory allocation.
https://github.com/lkl/linux/pull/487/files#diff-65cbb8e45dedadd081421ddc9002d12fR17

Reply:
LGTM, asking to @tavip for more comments if any.

Reply:
Could you please elaborate why is page_alloc needed? The only user I see is in bootmem_init and there we already make sure that we skip host memory until it is paged aligned.

Reply:
Good point! In this scenario (with the vfio backend), we can just use memory allocated by malloc().
Meanwhile, in other cases, I believe lkl should provide alternative memory allocation interface other than host_ops->alloc() for PCI DMA. For example, the architecture without IOMMU requires physically contiguous pages for bootmem, which is impossible to allocate with malloc(). Of course developers can override host_ops->alloc() for this, but it turns out inefficient memory consuming because host_ops->alloc() always has to return pages. This is why I proposed pci_ops->alloc() at the first commit.

However, your comment made me realize that the name "page_alloc" leads misunderstanding easily. What we need here is not a page allocation, but allocating memory suited for the PCI backend driver. I'm considering how I should fix and will apply that in a couple of days.

Reply:
Got it, thanks for the detailed explanation. pci_ops->alloc() sounds good (maybe alter we can convert it into a host dma_alloc if there are other users).

One thing that is no clear to me, is who will use this memory? I don't think we need to allocate bootmem with it, perhaps just internal PCI buffers?

Reply:
I totally agree that allocating the whole lkl memory with dma_alloc is not a sophisticated way.
However, linux components assume that the memory allocated by kmalloc() or other memory allocating functions is DMA-capable.

Documentation/DMA-API-HOWTO.txt says
> If you acquired your memory via the page allocator
> (i.e. __get_free_page*()) or the generic memory allocators
> (i.e. kmalloc() or kmem_cache_alloc()) then you may DMA to/from
> that memory using the addresses returned from those routines.

This makes things difficult, because it means that dma-capable memory allocators are often not called. (If it did, we could just hook dma_alloc() and return DMA-capable memory.)

There are some ways to solve this, but none of them are workable at modern device drivers such as NIC and NVMe.

The solutions I examined (and discarded) are 
- Using Zones to specify limited memory ranges for DMAs. This is for devices with limited address spaces, but does not work with the NVMe driver, which requires full address space. 
- bounce buffers. PCI drivers for some ARM SoCs implements this. Does not work as well, because the NVMe driver requires coherent memory and does not invoke DMA sync operation.

So, IMO, the only way to implement this PR is to make whole bootmem DMA-capable, even though this seems ugly....